### ISSUE TITLE
如何将ema/__0_0.distcp、ema/__1_0.distcp合并回ema.safetensors以适配inferencer.py进行推理

### ISSUE THREAD

**Issue by 13270285331** [ByteDance-Seed/Bagel#220]
https://github.com/ByteDance-Seed/Bagel/issues/139
您好，我在使用fsdp_save_fsdp_ckpt()生成权重文件ema/__0_0.distcp、ema/__1_0.distcp时，尝试将其转回ema.safetensors的形式，以适配inferencer.py进行推理，但一直未能成功，您空了能帮忙解决一下该问题嘛？
fsdp_save_fsdp_ckpt()生成权重文件格式如下：

<img width="1840" height="444" alt="Image" src="https://github.com/user-attachments/assets/0efff271-57cc-45b5-9501-23bda5757b02" />

**Comment by yangshurong** [comment]
import argparse
import torch.distributed.checkpoint as dist_cp
import torch.distributed.checkpoint.format_utils as dist_cp_format_utils
from safetensors.torch import save_file, load_file
state_dict = {}
dist_cp_format_utils._load_state_dict(
        state_dict,
        storage_reader=dist_cp.FileSystemReader(args.ckpt_dir),
        planner=dist_cp_format_utils._EmptyStateDictLoadPlanner(),
        no_dist=True,
    )
if len(state_dict.keys()) == 1:
        state_dict = state_dict[list(state_dict)[0]]
    save_file(state_dict, args.save_path)

你可以参考我的code

**Comment by Elijah123463** [comment]
> import argparse import torch.distributed.checkpoint as dist_cp import torch.distributed.checkpoint.format_utils as dist_cp_format_utils from safetensors.torch import save_file, load_file state_dict = {} dist_cp_format_utils._load_state_dict( state_dict, storage_reader=dist_cp.FileSystemReader(args.ckpt_dir), planner=dist_cp_format_utils._EmptyStateDictLoadPlanner(), no_dist=True, ) if len(state_dict.keys()) == 1: state_dict = state_dict[list(state_dict)[0]] save_file(state_dict, args.save_path)
> 
> 你可以参考我的code

好的，万分感谢🙏

我目前也写了一份，不过比较复杂：

1. 权重合并：

import os
import inspect
import logging
from copy import deepcopy
from dataclasses import dataclass, field

import torch
from safetensors.torch import save_file
from transformers import HfArgumentParser

from modeling.autoencoder import load_ae
from modeling.bagel import (
    BagelConfig, Bagel,
    Qwen2Config, Qwen2ForCausalLM,
    SiglipVisionConfig, SiglipVisionModel
)
from modeling.qwen2 import Qwen2Tokenizer
from data.data_utils import add_special_tokens


@dataclass
class ModelArguments:
    model_path: str = field(default="./BAGEL-7B-MoT/weights")
    llm_path: str = field(default="hf/Qwen2.5-0.5B-Instruct/")
    layer_module: str = field(default="Qwen2MoTDecoderLayer")
    llm_qk_norm: bool = field(default=True)
    tie_word_embeddings: bool = field(default=False)
    vae_path: str = field(default="flux/vae/ae.safetensors")
    vit_path: str = field(default="hf/siglip-so400m-14-980-flash-attn2-navit/")
    max_latent_size: int = field(default=32)      # <== 运行时用 --max_latent_size 64 覆盖
    latent_patch_size: int = field(default=2)
    vit_patch_size: int = field(default=14)
    vit_max_num_patch_per_side: int = field(default=70)
    connector_act: str = field(default="gelu_pytorch_tanh")
    interpolate_pos: bool = field(default=False)
    vit_select_layer: int = field(default=-2)
    vit_rope: bool = field(default=False)

@dataclass
class TrainingArguments:
    visual_gen: bool = field(default=True)
    visual_und: bool = field(default=True)
    freeze_und: bool = field(default=False)
    finetune_from_hf: bool = field(default=False)

@dataclass
class ExportArguments:
    ckpt_dir: str = field(metadata={"help": "step 目录（内含 ema/ 的 .distcp 分片）"})
    output: str = field(metadata={"help": "输出 ema.safetensors 路径"})
    pop_pos_embeds: bool = field(default=False)


def setup_logger():
    logger = logging.getLogger("merge_ema_distcp")
    logger.setLevel(logging.INFO)
    h = logging.StreamHandler()
    h.setFormatter(logging.Formatter("[merge] %(asctime)s - %(levelname)s - %(message)s"))
    logger.addHandler(h)
    return logger


def build_bagel(model_args: ModelArguments, train_args: TrainingArguments):
    if train_args.finetune_from_hf:
        llm_config = Qwen2Config.from_json_file(os.path.join(model_args.model_path, "llm_config.json"))
    else:
        llm_config = Qwen2Config.from_pretrained(model_args.llm_path)
    llm_config.layer_module = model_args.layer_module
    llm_config.qk_norm = model_args.llm_qk_norm
    llm_config.tie_word_embeddings = model_args.tie_word_embeddings
    llm_config.freeze_und = train_args.freeze_und

    language_model = (Qwen2ForCausalLM(llm_config)
                      if train_args.finetune_from_hf
                      else Qwen2ForCausalLM.from_pretrained(model_args.llm_path, config=llm_config))

    vit_config = None
    vit_model = None
    if train_args.visual_und:
        if train_args.finetune_from_hf:
            vit_config = SiglipVisionConfig.from_json_file(os.path.join(model_args.model_path, "vit_config.json"))
        else:
            vit_config = SiglipVisionConfig.from_pretrained(model_args.vit_path)
        vit_config.num_hidden_layers = vit_config.num_hidden_layers + 1 + model_args.vit_select_layer
        vit_config.rope = model_args.vit_rope

        vit_model = (SiglipVisionModel(vit_config)
                     if train_args.finetune_from_hf
                     else SiglipVisionModel.from_pretrained(model_args.vit_path, config=vit_config))


    vae_config = None
    if train_args.visual_gen:
        _, vae_config = load_ae(
            local_path=os.path.join(model_args.model_path, "ae.safetensors")
            if train_args.finetune_from_hf else model_args.vae_path
        )


    cfg = BagelConfig(
        visual_gen=train_args.visual_gen,
        visual_und=train_args.visual_und,
        llm_config=llm_config,
        vit_config=vit_config if train_args.visual_und else None,
        vae_config=vae_config if train_args.visual_gen else None,
        latent_patch_size=model_args.latent_patch_size,
        max_latent_size=model_args.max_latent_size,
        vit_max_num_patch_per_side=model_args.vit_max_num_patch_per_side,
        connector_act=model_args.connector_act,
        interpolate_pos=model_args.interpolate_pos,
        timestep_shift=1.0,
    )
    model = Bagel(language_model, vit_model if train_args.visual_und else None, cfg)


    if train_args.visual_und:
        model.vit_model.vision_model.embeddings.convert_conv2d_to_linear(vit_config)

    return model

def prepare_tokenizer_and_resize(model: Bagel, model_args: ModelArguments, train_args: TrainingArguments, logger: logging.Logger):

    tok_from = (model_args.model_path if train_args.finetune_from_hf else model_args.llm_path)
    tokenizer = Qwen2Tokenizer.from_pretrained(tok_from)

    tokenizer, new_token_ids, num_new_tokens = add_special_tokens(tokenizer)
    if num_new_tokens > 0:
        model.language_model.resize_token_embeddings(len(tokenizer))
        model.config.llm_config.vocab_size = len(tokenizer)
        model.language_model.config.vocab_size = len(tokenizer)
        logger.info(f"Resized token embeddings to {len(tokenizer)} (added {num_new_tokens} special tokens).")


def _init_single_process_group():
    import torch.distributed as dist
    if dist.is_initialized():
        return
    os.environ.setdefault("MASTER_ADDR", "127.0.0.1")
    os.environ.setdefault("MASTER_PORT", "29521")
    os.environ.setdefault("RANK", "0")
    os.environ.setdefault("WORLD_SIZE", "1")
    dist.init_process_group(backend="gloo", rank=0, world_size=1)

def _load_distcp_into_state_dict(target_sd: dict, ema_dir: str, logger: logging.Logger):
    from torch.distributed.checkpoint import FileSystemReader
    reader = FileSystemReader(ema_dir)

    try:
        from torch.distributed.checkpoint import load as tdc_load
        params = set(inspect.signature(tdc_load).parameters.keys())
        if "no_dist" in params:
            logger.info("Using TDC load(..., no_dist=True).")
            tdc_load(target_sd, storage_reader=reader, no_dist=True)
            return
        else:
            logger.info("TDC load() without no_dist -> init single-process group.")
            _init_single_process_group()
            tdc_load(target_sd, storage_reader=reader)
            return
    except Exception as e:
        logger.warning(f"TDC load() failed with: {e}. Falling back to load_state_dict().")

    from torch.distributed.checkpoint import load_state_dict as tdc_load_state_dict
    _init_single_process_group()
    tdc_load_state_dict(target_sd, reader)


def main():
    parser = HfArgumentParser((ModelArguments, TrainingArguments, ExportArguments))
    model_args, train_args, export_args = parser.parse_args_into_dataclasses()
    logger = setup_logger()

    ema_dir = os.path.join(export_args.ckpt_dir, "ema")
    if not os.path.isdir(ema_dir):
        raise FileNotFoundError(f"未找到分片目录：{ema_dir}")
    logger.info(f"Reading shards from: {ema_dir}")

    model = build_bagel(model_args, train_args)
    model.eval()

    prepare_tokenizer_and_resize(model, model_args, train_args, logger)

    target_sd = deepcopy(model.state_dict())  
    _load_distcp_into_state_dict(target_sd, ema_dir, logger)


2. 补全缺失权重

import argparse
import torch
from safetensors import safe_open
from safetensors.torch import save_file

def overlay(a_path, b_path, out_path, save_dtype="bf16"):
    out = {}
    with safe_open(a_path, framework="pt", device="cpu") as fa, \
         safe_open(b_path, framework="pt", device="cpu") as fb:

        keys_a = set(fa.keys())
        keys_b = list(fb.keys())

        for k in keys_b:
            t = fa.get_tensor(k) if k in keys_a else fb.get_tensor(k)
            if t.is_floating_point():
                if save_dtype.lower() in ["bf16", "bfloat16"]:
                    t = t.to(torch.bfloat16)
                elif save_dtype.lower() in ["fp32", "float32"]:
                    t = t.to(torch.float32)
            out[k] = t

    save_file(out, out_path)
    print(f"Saved merged file: {out_path} (dtype={save_dtype})")
    print(f"keys: {len(out)}; example: {list(out.keys())[:10]}")

if __name__ == "__main__":
    ap = argparse.ArgumentParser()
    ap.add_argument("--a", required=True, help="A: 你合并后的文件（覆盖来源）")
    ap.add_argument("--b", required=True, help="B: 原始文件（作为基底，键更全）")
    ap.add_argument("--out", required=True, help="输出路径")
    ap.add_argument("--save_dtype", default="bf16", choices=["bf16","bfloat16","fp32","float32"])
    args = ap.parse_args()
    overlay(args.a, args.b, args.out, args.save_dtype)

感谢您的分享，我晚会试试的～
